rfcs/20231129-tuning-api/README.md

```diff
@@ -0,0 +1,134 @@
+
+### Option 1- (Recommended)
+Basic Usage Scenario
+```c
+create_primitive(prim); 
+dnnl_set_tune(true);   
+execute(prim);  //tuning happens here
+dnnl_set_tune(false); 
+create_primitive(prim); //primitive is created with config selected from tuning
+execute(prim); //normal execution
+```
+Tuning happens during one call of the execute function on the primitive. 
+Subsequent execute calls on the same primitive will not result in re-tuning the primitive.  
+Tuned results will be stored in a primitive implementation specific lookup table that will be referenced  
+when the primitive is (re)created. (Some gpu implementations such as conv and batchnorm already use lookup tables.)  
+Tuning will happen under a cold cache mode and will be limited to max_nconfigs.  
+Primitive cache entry for the primitive(s) being tuned will be updated to point to the tuned implementation when tuning is complete.  
+Kernel cache entries will be unmodified for now, but can be modified if we want to enable tuning of resuable kernels later.  
+If the user wants to persist the tuned configs between sessions, the lookup tables can optionally be written to files.  
+
+### Option 2
+Basic Usage Scenario
+```c
+dnnl_set_tune(true);   
+create_primitive(prim); // tuning happens here
+dnnl_set_tune(false); 
+execute(prim); //normal execution with tuned implementation
+```
+Unlike the first option, primitive does not have to be recreated afterward.
+However, oneDNN will have to allocate and initialize all memory needed for execution internally during creation.
+This adds additional complexity to the implementation, potentially high memory consumption
+and need for an optimized data filling routine.
+
+Since frameworks seem ok with first option, would recommend Option 1.
+
+### Implementation Details
+The following structure will be added to primitive_attr_t.
+```c
+struct tune_info_t {
+    void set_tune_iter(int i); // set configuration to try
+    void set_tune_profile(int i, double time); //record min time for ith configuration 
+    enum tune_status_t { searching /*default*/, finalize }; 
+    int iter = -1; //which configuration to try
+    std::vector<double> iter_profile_time; // measured time for ith configuration
+    int max_iters = 1; //max number of iters to try obtained by querying implementation
+    tune_status_t tune_status = searching; //search or finalize status
+};
+```
+During the primitive execute call, will query the implementation for the number of configs it has via
+`query::tune_nconfigs`.  For each config it will create the primitive, execute it 5 times, and record the min
+time in the tune_info_t structure. In the case where number of configurations to try is greater than 40 it will stop.
+It will then recreate the primitive with tune_status set to finalize. During this call the config with the best 
+performance will be stored in a lookup table managed by the primitive and primitive cache will be updated to point
```

Comment:
If this mechanism is to be extended to many other primitives - it may make sense to manage tuned parameters at a higher level in a generic way. Otherwise every primitive would have to implement lookup tables, logic for the finalization step, etc. But this is an implementation detail, and can be adjusted in the future if needed without API changes.